File: templates/python-mcp/src/datasheet_mcp/uris.py
"""Canonical `datasheet://` resource identities."""

from __future__ import annotations

SCHEME = "datasheet"
REPORTS_URI = "datasheet://reports"
WORKBENCH_APP_URI = "ui://datasheet/workbench.html"
USAGE_ROOT_URI = "datasheet://usage"
USAGE_TASK_TEMPLATE = "datasheet://usage/task/{task_id}"
ARTIFACT_TEMPLATE = "datasheet://artifact/{artifact_id}"
DOCS_URI = "datasheet://docs"
CONTRACT_URI = "datasheet://contract"

_USAGE_TASK_PREFIX = "datasheet://usage/task/"
_ARTIFACT_PREFIX = "datasheet://artifact/"
_DOC_PREFIX = "datasheet://docs/"


def usage_task_uri(task_id: str) -> str:
    return f"{_USAGE_TASK_PREFIX}{task_id}"


def artifact_uri(artifact_id: str) -> str:
    return f"{_ARTIFACT_PREFIX}{artifact_id}"


def parse_usage_task_uri(uri: str) -> str | None:
    if uri.startswith(_USAGE_TASK_PREFIX):
        task_id = uri[len(_USAGE_TASK_PREFIX) :]
        return task_id or None
    return None


def parse_artifact_uri(uri: str) -> str | None:
    if uri.startswith(_ARTIFACT_PREFIX):
        artifact_id = uri[len(_ARTIFACT_PREFIX) :]
        return artifact_id or None
    return None


def doc_uri(doc_id: str) -> str:
    return f"{_DOC_PREFIX}{doc_id}"


def parse_doc_uri(uri: str) -> str | None:
    if uri.startswith(_DOC_PREFIX):
        doc_id = uri[len(_DOC_PREFIX) :]
        return doc_id or None
    return None

File: servers/duckdb-mcp/src/uris.rs
use veoveo_mcp_contract::{ArtifactId, ServerResourceUris};

pub const DBS_ROOT_URI: &str = "duckdb://dbs";
pub const WORKBENCH_APP_URI: &str = "ui://duckdb/workbench.html";
pub const DB_TEMPLATE: &str = "duckdb://db/{db_id}";
pub const ARTIFACT_TEMPLATE: &str = "duckdb://artifact/{artifact_id}";
pub const USAGE_ROOT_URI: &str = "duckdb://usage";
pub const USAGE_TASK_TEMPLATE: &str = "duckdb://usage/task/{task_id}";

/// Well-known surface roots (contract C18, C19). These literals must match
/// `ServerResourceUris::new("duckdb")`; a unit test below pins the
/// equivalence.
pub const DOCS_URI: &str = "duckdb://docs";
pub const CONTRACT_URI: &str = "duckdb://contract";
pub const DOC_TEMPLATE: &str = "duckdb://docs/{doc_id}";

fn duckdb_uris() -> ServerResourceUris {
    ServerResourceUris::new("duckdb")
}

pub fn db_uri(db_id: &str) -> String {
    format!("duckdb://db/{db_id}")
}

pub fn parse_db_uri(uri: &str) -> Option<&str> {
    let rest = uri.strip_prefix("duckdb://db/")?;
    if rest.is_empty() || rest.contains('/') {
        None
    } else {
        Some(rest)
    }
}

pub fn doc_uri(doc_id: &str) -> String {
    duckdb_uris().doc_uri(doc_id)
}

pub fn parse_doc_uri(uri: &str) -> Option<&str> {
    veoveo_mcp_contract::parse_server_doc_uri("duckdb", uri)
}

pub fn artifact_uri(artifact_id: ArtifactId) -> String {
    duckdb_uris().artifact_uri(artifact_id)
}

pub fn usage_task_uri(task_id: &str) -> String {
    duckdb_uris().usage_task_uri(task_id)
}

pub fn parse_artifact_uri(uri: &str) -> Option<ArtifactId> {
    duckdb_uris().parse_artifact_uri(uri)
}

pub fn parse_usage_task_uri(uri: &str) -> Option<&str> {
    duckdb_uris().parse_usage_task_uri(uri)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn db_uri_round_trips() {
        let uri = db_uri("robot_metrics");
        assert_eq!(uri, "duckdb://db/robot_metrics");
        assert_eq!(parse_db_uri(&uri), Some("robot_metrics"));
        assert_eq!(parse_db_uri("duckdb://db/"), None);
        assert_eq!(parse_db_uri("duckdb://db/a/b"), None);
        assert_eq!(parse_db_uri("duckdb://dbs"), None);
    }

    #[test]
    fn well_known_uris_match_the_shared_contract_conventions() {
        let conventions = duckdb_uris();
        assert_eq!(DOCS_URI, conventions.docs_root_uri());
        assert_eq!(CONTRACT_URI, conventions.contract_uri());
        assert_eq!(DOC_TEMPLATE, conventions.doc_template());
        assert_eq!(doc_uri("agents"), conventions.doc_uri("agents"));
        assert_eq!(parse_doc_uri("duckdb://docs/agents"), Some("agents"));
        assert_eq!(parse_doc_uri("duckdb://docs"), None);
        assert_eq!(parse_doc_uri("duckdb://docs/agents/extra"), None);
    }

    #[test]
    fn artifact_uri_round_trips() {
        let artifact_id = ArtifactId::new();
        let uri = artifact_uri(artifact_id);
        assert_eq!(uri, format!("duckdb://artifact/{artifact_id}"));
        assert_eq!(parse_artifact_uri(&uri), Some(artifact_id));
    }
}
